staging/src/k8s.io/client-go/rest/with_retry.go

```diff
@@ -194,46 +198,18 @@ func (r *withRetry) IsNextRetry(ctx context.Context, restReq *Request, httpReq *
 	r.retryAfter.Wait = time.Duration(seconds) * time.Second
 	r.retryAfter.Reason = getRetryReason(r.attempts, seconds, resp, err)
 
-	if err := r.prepareForNextRetry(ctx, restReq); err != nil {
-		klog.V(4).Infof("Could not retry request - %v", err)
-		return false
-	}
-
 	return true
 }
 
-// prepareForNextRetry is responsible for carrying out operations that need
-// to be completed before the next retry is initiated:
-// - if the request context is already canceled there is no need to
-//   retry, the function will return ctx.Err().
-// - we need to seek to the beginning of the request body before we
-//   initiate the next retry, the function should return an error if
-//   it fails to do so.
-func (r *withRetry) prepareForNextRetry(ctx context.Context, request *Request) error {
-	if ctx.Err() != nil {
-		return ctx.Err()
-	}
-
-	// Ensure the response body is fully read and closed before
-	// we reconnect, so that we reuse the same TCP connection.
-	if seeker, ok := request.body.(io.Seeker); ok && request.body != nil {
-		if _, err := seeker.Seek(0, 0); err != nil {
-			return fmt.Errorf("can't Seek() back to beginning of body for %T", request)
-		}
-	}
-
-	klog.V(4).Infof("Got a Retry-After %s response for attempt %d to %v", r.retryAfter.Wait, r.retryAfter.Attempt, request.URL().String())
-	return nil
-}
-
 func (r *withRetry) Before(ctx context.Context, request *Request) error {
+	// If the request context is already canceled there
```

Comment:
I had a question about After for @tkashem here https://github.com/kubernetes/kubernetes/pull/108347/files#r838810333 - it's not clear to me what the value of After is at this point (may be something his refactoring to fix thread safety in Request will have to deal with)

Reply:
+1, I've been wondering the same. 
Would it maybe make sense to put reading and closing response body as part of `After()`? 
Seems intuitive in the flow of `Before` -> attempt -> `After`
and the caller doesn't have to handle it explicitly because the implementation would do it for them (and considering forgetting to implement it could lead to races again, maybe we could consider making reading and closing part of the implementation itself)

Reply:
> Would it maybe make sense to put reading and closing response body as part of After()?

it will be a little tricky to achieve that given that we have transform function, but definitely something we can look into during the follow-up refactor.

for now, we need `After` to update the backoff which needs the reply tuple (response, err) from the server.

Reply:
oh 100%, definitely part of a follow up!